homeassistant/components/hdmi_cec.py

```diff
@@ -320,38 +320,40 @@ def _start_cec(event):
 class CecDevice(Entity):
     """Representation of a HDMI CEC device entity."""
 
-    def __init__(self, hass: HomeAssistant, device, logical) -> None:
+    def __init__(self, device, logical) -> None:
         """Initialize the device."""
         self._device = device
-        self.hass = hass
         self._icon = None
         self._state = STATE_UNKNOWN
         self._logical_address = logical
         self.entity_id = "%s.%d" % (DOMAIN, self._logical_address)
-        device.set_update_callback(self._update)
 
     def update(self):
         """Update device status."""
-        self._update()
+        device = self._device
+        from pycec.const import STATUS_PLAY, STATUS_STOP, STATUS_STILL, \
+            POWER_OFF, POWER_ON
+        if device.power_status in [POWER_OFF, 3]:
+            self._state = STATE_OFF
+        elif device.status == STATUS_PLAY:
+            self._state = STATE_PLAYING
+        elif device.status == STATUS_STOP:
+            self._state = STATE_IDLE
+        elif device.status == STATUS_STILL:
+            self._state = STATE_PAUSED
+        elif device.power_status in [POWER_ON, 4]:
+            self._state = STATE_ON
+        else:
+            _LOGGER.warning("Unknown state: %d", device.power_status)
+
+    async def async_added_to_hass(self):
+        """Register HDMI callbacks after initialization."""
+        self._device.set_update_callback(self._update)
 
+    @callback
```

Comment:
Do we know that this callback will be called within the event loop? Otherwise it shouldn't be decorated with `@callback`.

Reply:
It uses `hass.loop` sometimes https://github.com/home-assistant/home-assistant/blob/c6fc6d8b051487a1c4e38e2ee43e99e972aeee56/homeassistant/components/hdmi_cec.py#L173-L183

Reply:
Callback decorator is removed as it called from 3rd party software. cc @balloob